qualcommax: ipq807x: Add ability to perform upgrade on current partition and remove OEM UBI volume for Linksys MX devices

testuser7 opened the pull request:
Dual firmware devices like Linksys MX4200, MX4300, MX5300 and MX8500 use separate rootfs partition.
This option helps to keep all configuration during upgrade.

1. Add new sysupgrade option to be able to perform upgrade on current partition for dual firmware devices:
  `-s           stay on current partition (for dual firmware devices)`

2. Update u-boot env variables only when changes are made.

3. Use `remove_oem_ubi_volume` function to remove OEM UBI volume before upgrade.
This allows to upgrade even if we have OEM firmware on the second partition.

testuser7 committed:
base-files: sysupgrade: Add new option -s to perform upgrade on current partition

Add new option to be able to perform upgrade on current partition for dual firmware devices:
  "-s           stay on current partition (for dual firmware devices)"

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14720
Signed-off-by: Robert Marko <robimarko@gmail.com>

testuser7 committed:
qualcommax: ipq807x: Add ability to perform upgrade on current partition for Linksys MX devices

Dual firmware devices like Linksys MX4200, MX4300, MX5300 and MX8500 use separate rootfs partition.
This option helps to keep all configuration during upgrade.

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14720
Signed-off-by: Robert Marko <robimarko@gmail.com>

testuser7 commented:
Is anyone interested in these changes?

NTheZone commented:
> Is anyone interested in these changes?

Yes

testuser7 commented:
> Yes. Include MX4300/MX4301

May be added after merge of this PR: https://github.com/openwrt/openwrt/pull/16070

cybrnook commented:
> Is anyone interested in these changes?

Yes, this would be great for dual partition systems. Much easier than having to switch back to alt partition to flash and then switch back.

cybrnook commented:
> > Is anyone interested in these changes?
> 
> Yes. Include MX4300/MX4301

@NTheZone if you pull this PR down as a patch and apply it on top of the PR adding support for the MX4300, functionality for the MX4300 works already today. I tested this last night and upgrading using -s from the cli works fine. I test flashed twice without issue.

OpenWRT-fanboy commented:
@megalow2k @cybrnook @NTheZone if you like it, you should approve it :)

NTheZone reviewed (approved):
Approving.  I cannot opine precisely on syntax and logic, necessarily, but it looks OK to me.

cybrnook reviewed (approved):
Worked fine on a MX4300, test flashed twice.

hnyman commented:
To me the logic looks faulty, as the line `CI_UBIPART="rootfs"` disappears from the decision tree and only alt_rootfs is used. That line should be in the last else block, I think.

musashino205 commented:
`CURR_PARTITION` looks like a partition index, `USE_CURR_PART` or something is better for me.

georgemoussalem reviewed (approved):
tested on ipq50xx plafform on 2 linksys devices. I do agree with the idea of changing the CURR_PARTITION variable name itself to more clearly indicate staying on the current partition for semantics, it's not a blocker to me though.

testuser7 commented:
> To me the logic looks faulty, as the line `CI_UBIPART="rootfs"` disappears from the decision tree and only alt_rootfs is used. That line should be in the last else block, I think.

`rootfs` is the default value for the `CI_UBIPART` variable and does not need to be specified again.

testuser7 commented:
> IMO, `CURR_PARTITION` looks like a partition index, `USE_CURR_PART` or something is better for me.

Corrected.

testuser7 commented:
Now I am confused which variable should we use if kernel is on separate partition?

```
# 'kernel' partition or UBI volume on NAND contains the kernel
CI_KERNPART="${CI_KERNPART:-kernel}"

# 'ubi' partition on NAND contains UBI
# There are also CI_KERN_UBIPART and CI_ROOT_UBIPART if kernel
# and rootfs are on separated UBIs.
CI_UBIPART="${CI_UBIPART:-ubi}"

# 'rootfs' UBI volume on NAND contains the rootfs
CI_ROOTPART="${CI_ROOTPART:-rootfs}"
```

musashino205 commented:
## raw kernel in mtd and rootfs in UBI

`CI_KERNPART` + `CI_UBIPART`

kernel ---> written to the mtd partition `"$CI_KERNPART" (default: "kernel")`
rootfs ---> written to the UBI volume `"$CI_ROOTPART" (default: "rootfs")` in the mtd partition `"$CI_UBIPART" (default: "ubi")`
rootfs_data ---> create a UBI volume `"rootfs_data"` in the mtd partition `"$CI_UBIPART"`

## kernel and rootfs in the same UBI

`CI_KERNPART` + `CI_UBIPART`

kernel ---> written to the UBI volume `"$CI_KERNPART"` in  the mtd partition `"$CI_UBIPART"`
rootfs ---> written to the UBI volume `"$CI_ROOTPART"` in  the mtd partition `"$CI_UBIPART"`
rootfs_data ---> create a UBI volume `"rootfs_data"` in  the mtd partition `"$CI_UBIPART"`

## kernel and rootfs in the different UBI

`CI_KERN_UBIPART` + `CI_ROOT_UBIPART` + `CI_KERNPART` + `CI_ROOTPART`

kernel ---> written to the UBI volume `"$CI_KERNPART"` in the mtd partition `"$CI_KERN_UBIPART"`
rootfs ---> written to the UBI volume `"$CI_ROOTPART"` in the mtd partition `"$CI_ROOT_UBIPART"`
rootfs_data ---> create a UBI volume `"rootfs_data"` in the mtd partition `"$CI_ROOT_UBIPART"`

testuser7 committed:
qualcommax: ipq807x: Update u-boot env only when changes for Linksys MX devices

Update u-boot env variables only when changes are made.
Applies to Linksys MX4200/MX4300/MX5300/MX8500.

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14720
Signed-off-by: Robert Marko <robimarko@gmail.com>

testuser7 commented:
@musashino205 The script probably worked because of this: https://github.com/openwrt/openwrt/blob/main/package/base-files/files/lib/upgrade/nand.sh#L396

musashino205 commented:
> @musashino205 The script probably worked because of this: https://github.com/openwrt/openwrt/blob/main/package/base-files/files/lib/upgrade/nand.sh#L396

Ah, I didn't noticed it. But IMO, `CI_UBIPART` should be set explicitly instead of fallback.

robe-zhang commented:
> Is anyone interested in these changes?

Does it upgrade current partition and reboot to current partition, not another partition before?
And how to upgrade on current partition?

testuser7 commented:
> Does it upgrade current partition and reboot to current partition, not another partition before?

Yes, that is exactly the purpose of this PR.

testuser7 committed:
qualcommax: ipq807x: Remove OEM UBI volume before upgrade for Linksys MX devices

Use 'remove_oem_ubi_volume' function to remove OEM UBI volume before upgrade.
This allows to upgrade even if we have OEM firmware on the second partition.
Applies to Linksys MX4200/MX4300/MX5300/MX8500.

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/14720
Signed-off-by: Robert Marko <robimarko@gmail.com>

testuser7 commented:
> Yes. Include MX4300/MX4301

Now MX4300 is supported.

testuser7 commented:
@robimarko what do you think about using the `remove_oem_ubi_volume` function?

robimarko commented:
Thanks! Rebased on top of main and merged!
